src/NeoLocalPlanner.cpp

```diff
@@ -771,49 +818,54 @@ void NeoLocalPlanner::initialize(std::string name, tf2_ros::Buffer* tf, costmap_
-	m_limits.min_vel_theta = 		private_nh.param<double>("min_rot_vel", 0.1);
-	m_limits.max_vel_theta = 		private_nh.param<double>("max_rot_vel", 0.5);
-	m_limits.min_vel_trans = 		private_nh.param<double>("min_trans_vel", 0.1);
-	m_limits.max_vel_trans = 		private_nh.param<double>("max_trans_vel", m_limits.max_vel_x);
-	m_limits.theta_stopped_vel = 	private_nh.param<double>("rot_stopped_vel", 0.5 * m_limits.min_vel_theta);
-	m_limits.trans_stopped_vel = 	private_nh.param<double>("trans_stopped_vel", 0.5 * m_limits.min_vel_trans);
-	m_limits.yaw_goal_tolerance = 	private_nh.param<double>("yaw_goal_tolerance", 0.02);
-	m_limits.xy_goal_tolerance = 	private_nh.param<double>("xy_goal_tolerance", 0.1);
-
-	m_differential_drive = 	private_nh.param<bool>("differential_drive", true);
-	m_constrain_final = 	private_nh.param<bool>("constrain_final", false);
-	m_goal_tune_time = 		private_nh.param<double>("goal_tune_time", 0.5);
-	m_lookahead_time = 		private_nh.param<double>("lookahead_time", 0.2);
-	m_lookahead_dist = 		private_nh.param<double>("lookahead_dist", 0.5);
-	m_start_yaw_error = 	private_nh.param<double>("start_yaw_error", 0.2);
-	m_pos_x_gain = 			private_nh.param<double>("pos_x_gain", 1);
-	m_pos_y_gain = 			private_nh.param<double>("pos_y_gain", 1);
-	m_pos_y_yaw_gain = 		private_nh.param<double>("pos_y_yaw_gain", 1);
-	m_yaw_gain = 			private_nh.param<double>("yaw_gain", 1);
-	m_static_yaw_gain = 	private_nh.param<double>("static_yaw_gain", 3);
-	m_cost_x_gain = 		private_nh.param<double>("cost_x_gain", 0.1);
-	m_cost_y_gain = 		private_nh.param<double>("cost_y_gain", 0.1);
-	m_cost_y_yaw_gain = 	private_nh.param<double>("cost_y_yaw_gain", 0.1);
-	m_cost_y_lookahead_dist = 	private_nh.param<double>("cost_y_lookahead_dist", 0);
-	m_cost_y_lookahead_time = 	private_nh.param<double>("cost_y_lookahead_time", 1);
-	m_cost_yaw_gain = 		private_nh.param<double>("cost_yaw_gain", 1);
-	m_low_pass_gain = 		private_nh.param<double>("low_pass_gain", 0.5);
-	m_max_cost = 			private_nh.param<double>("max_cost", 0.9);
-	m_max_curve_vel = 		private_nh.param<double>("max_curve_vel", 0.2);
-	m_max_goal_dist = 		private_nh.param<double>("max_goal_dist", 0.5);
-	m_max_backup_dist = 	private_nh.param<double>("max_backup_dist", m_differential_drive ? 0.1 : 0.0);
-	m_min_stop_dist = 		private_nh.param<double>("min_stop_dist", 0.5);
-	m_emergency_acc_lim_x = private_nh.param<double>("emergency_acc_lim_x", m_limits.acc_lim_x * 4);
-	m_enable_software_stop = private_nh.param<bool>("enable_software_stop", true);
-	m_allow_reversing = private_nh.param<bool>("allow_reversing", false);
+	dsrv_ = new dynamic_reconfigure::Server<NeoLocalPlannerConfig>(private_nh);
+    dynamic_reconfigure::Server<NeoLocalPlannerConfig>::CallbackType cb =
+        boost::bind(&NeoLocalPlanner::reconfigureCB, this, _1, _2);
+    dsrv_->setCallback(cb);
+
+	// m_limits.acc_lim_x = 			private_nh.param<double>("acc_lim_x", 0.5);
```

Comment:
Actually the default values are in the NeoPlanner.cfg under /cfg folder. But if you have a yaml config file and load it using rosparam then it would default to those values. Dynamic reconfigure uses ros param as default parameters.

Reply:
>  But if you have a yaml config file and load it using rosparam then it would default to those values.

Then how would the node know about the parameters from the yaml ? 

We need to keep in mind that `rosparam` command does not invoke the dynamic reconfigure callback. Therefore it is essential to have same set of initial / default values to the cfg and yaml files.

Reply:
I think it's easiest to try it out for yourself. If you load in a yaml file using rosparam example:
 `<rosparam file="$(find tc_motion_planning)/config/neo_local_planner_params.yaml"
                command="load"/>` 

and if you have `max_vel_trans:  3.0` in that param file. Then when you launch move_base, you will notice that in dynamic reconfigure, the `max_vel_trans` is `3.0` instead of the default that I defined in NeoPlanner.cfg line 19 `gen.add("max_trans_vel", double_t, 0, "The absolute value of the maximum translational velocity for the robot in m/s", 0.5, 0.0, 3.0)` which is `0.5`.

If you don't have a parameter in the yaml file, then ROS will just default to using the default value in the cfg file. Let me know if you obtain the same result after trying the above.

Reply:
I understand the point of assigning the default value to a parameter through CFG file. To elaborate a bit, does the rosparam load overwrites the default parameters in the parameter server through rosparam load command in the launch file? Yes! But how does the ros node know the changes made to the parameter server and reflects it in the calculation that it makes?

Reply:
The default dynamic reconfigure server variables will be overwritten by the yaml file. The node will then extract those parameters when it instantiates the dynamic reconfigure server and sets up the callback function in line 821: `dsrv_ = new dynamic_reconfigure::Server<NeoLocalPlannerConfig>(private_nh);.` When the call back is triggered, those **modified** parameters will then be extracted into the node from the callback in line 170: `void NeoLocalPlanner::reconfigureCB(NeoLocalPlannerConfig& config,uint32_t level) {`

Reply:
Cool.. as long as you have tested it 😃

Reply:
Okay.. I had some time now to test it as well.. it works. But I'm not really a fan of initialising the variable using the dynamic reconfigure. It feels like a work around. The task of the dynamic reconfigure is to set dynamic parameters and not initialising the parameters. Therefore please bring back the parameter initialisation just for the sake of intent in which each function needs to be used.